podman_compose.py

```diff
@@ -1156,17 +1158,29 @@ def flat_deps(services, with_extends=False):
 
 
 class Podman:
-    def __init__(self, compose, podman_path="podman", dry_run=False):
+    def __init__(self, compose, podman_path="podman", dry_run=False, semaphore: asyncio.Semaphore = asyncio.Semaphore(sys.maxsize)):
         self.compose = compose
         self.podman_path = podman_path
         self.dry_run = dry_run
+        self.semaphore = semaphore
+
+    async def output(self, podman_args, cmd="", cmd_args=None):
+        async with self.semaphore:
```

Comment:
why do you need semaphore here? it's supposed to run on a different subprocess. it seems semaphore is not needed.

Reply:
It's an asyncio semaphore. It's to support limiting the number of parallel operations. I'm fairly sure this is compatible with the way the v2 docker compose plugin works. Where it's parallel by default (the defalut allows maxint number of parallel operations, but you can limit it with `--parallel <n>` or `COMPOSE_PARALLEL_LIMIT`.

I followed the concerns around this https://github.com/docker/compose/issues/8226 and other issues

Reply:
I noticed that in many places we can replace `.run` with `.exec` which will give us simpler code and better error reporting or return code. for example

```
await compose.podman.run(["--version"], "", [])
```

should we remove `log_formatter` as it sometimes misses last line and since we have `podman logs --color`
it will also make code simpler
I think we should change the code to create containers then start them detached then run `podman logs --color` what do you think?

I see that `podman.output()` is only used internally so it does not need to be semaphored.

Reply:
Well it's not the number of CPUs. This is entirely single threaded now, so it will only ever use 1 cpu. It's just the number of concurrent tasks that are allowed, and docker compose's default is unbounded, which I've just replicated.

I was just trying to keep the changes as small as possible rather, so if you think cleaning up the run/exec makes sense that works.

Wouldn't starting the containers detached complicate this though? It would mean that if compose gets killed via sigkill, it won't be able to clean up the containers it started. I'm not an expert here but I would think that could likely lead to issues for users.

Reply:
> Well it's not the number of CPUs. This is entirely single threaded now, so it will only ever use 1 cpu. It's just the number of concurrent tasks that are allowed, and docker compose's default is unbounded, which I've just replicated.

it's in deed, we are forking a subprocess like `podman build` or something like that
even if we not, it's a good heuristic indicator of how much the machine could handle.

> I was just trying to keep the changes as small as possible rather

I've merged it, but I'll need someone like you to become co-maintainer and help me refactor the code.

>  so if you think cleaning up the run/exec makes sense that works.

I'll open an issue with the points in my mind. and your PRs are welcome.

> Wouldn't starting the containers detached complicate this though? 

it made CTRL+C works. I know it will break things, that's why I created a different branch.